services/oauth/v2/oauth.go

```diff
@@ -0,0 +1,396 @@
+package v2
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	rudderSync "github.com/rudderlabs/rudder-go-kit/sync"
+	obskit "github.com/rudderlabs/rudder-observability-kit/go/labels"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	router_utils "github.com/rudderlabs/rudder-server/router/utils"
+)
+
+func WithCache(cache Cache) func(*OAuthHandler) {
+	return func(h *OAuthHandler) {
+		h.Cache = cache
+	}
+}
+
+func WithLocker(lock *rudderSync.PartitionRWLocker) func(*OAuthHandler) {
+	return func(h *OAuthHandler) {
+		h.CacheMutex = lock
+	}
+}
+
+func WithExpirationTimeDiff(expirationTimeDiff time.Duration) func(*OAuthHandler) {
+	return func(h *OAuthHandler) {
+		h.ExpirationTimeDiff = expirationTimeDiff
+	}
+}
```

Comment:
We have functional options for these. We can probably make in unexported.

```suggestion
func WithCache(cache Cache) func(*OAuthHandler) {
	return func(h *OAuthHandler) {
		h.cache = cache
	}
}

func WithLocker(lock *rudderSync.PartitionRWLocker) func(*OAuthHandler) {
	return func(h *OAuthHandler) {
		h.cacheMutex = lock
	}
}

func WithExpirationTimeDiff(expirationTimeDiff time.Duration) func(*OAuthHandler) {
	return func(h *OAuthHandler) {
		h.expirationTimeDiff = expirationTimeDiff
	}
}
```